k2/csrc/context_inl.cuh

```diff
@@ -0,0 +1,182 @@
+      {MemcpyDeviceToHost, cudaMemcpyDeviceToHost},
+      {MemcpyDeviceToDevice, cudaMemcpyDeviceToDevice}};
+  auto it = copy_kind_mappings.find(kind);
+  K2_CHECK_NE(it, copy_kind_mappings.end());
+  cudaMemcpy(dst, src, count, it->second);
+}
+
+template <typename T1, typename T2>
+bool IsCompatible(const T1 &t1, const T2 &t2) {
+  // suppose both T1 and T2 have member method `Context`
+  return t1.Context()->IsCompatible(*t2.Context());
+}
+
+template <typename T>
+ContextPtr GetContext(T &t) {
+  // suppose T has member method `Context`
+  ContextPtr c;
+  t.To(c);
+  return c;
+}
+
+template <typename First, typename... Rest>
+ContextPtr GetContext(First &first, Rest &... rest) {
+  ContextPtr ans1 = GetContext(first), ans2 = GetContext(rest...);
+  K2_DCHECK(ans1->IsCompatible(*ans2)) << "Contexts are not compatible";
+  return ans1;
+}
+
+/*
+  Convenience wrapper for NewRegion() that takes the context from a provided
+  region.
+ */
+inline std::shared_ptr<Region> NewRegion(Region &region, std::size_t num_bytes) {
+  return NewRegion(region.context, num_bytes);
+}
+
+// Objects from k2 generally have a Context() method, so this template
+// will work to get the device-type for pretty arbitrary objects.
+template <typename T>
+inline DeviceType DeviceOf(const T &t) {
+  return t.Context()->GetDeviceType();
+}
+
+/* Eval() will evaluate lambda(i) for 0 <= i < n, on the appropriate
+   device (CPU or GPU). */
+template <typename LambdaT>
+void Eval(cudaStream_t stream, int32_t n, LambdaT &lambda) {
+  if (n <= 0) return;  // actually it would be an error if n < 0.
+  if (stream == kCudaStreamInvalid) {
+    // TODO: if n is very large, we'll eventually support running this with
+    // multiple threads.
+    for (int32_t i = 0; i < n; ++i) {
+      lambda(i);
+    }
+  } else {
+#ifdef __CUDA_ARCH__
+    int32_t block_size = 256;
+    int32_t grid_size = NumBlocks(n, block_size);
+    eval_lambda<LambdaT><<<grid_size, block_size, 0, stream>>>(n, lambda);
+    K2_DCHECK_CUDA_ERROR(cudaGetLastError());
+#endif
+  }
+}
+
+/*
+  This is a form of Eval() where the lambda takes  two arguments.
+
+  Eval2() will evaluate lambda(i, j) for 0 <= i < m and 0 <= j < n,
+  on the appropriate  device (CPU or GPU).  The second index, n,
+  is supposed to be the faster-varying one, the index for which
+  threads in the same warp will tend to have different values.
+  (Of course this doesn't affect the semantics of the operation).
+*/
+template <typename LambdaT>
+void Eval2(cudaStream_t stream, int32_t m, int32_t n, LambdaT &lambda) {
+  if (m <= 0 || n <= 0)
+    return;  // actually it would be an error if m < 0 or n < 0.
+  if (stream == kCudaStreamInvalid) {
+    // TODO: if n is very large, we'll eventually support running this with
+    // multiple threads.
+    for (int32_t i = 0; i < m; ++i) {
+      for (int32_t j = 0; j < n; ++j) {
+        lambda(i, j);
+      }
+    }
+  } else {
+#ifdef __CUDA_ARCH__
```

Comment:
May not use this macro. @danpovey defined `kCudaStreamInvalid` to distinguish CPU stream from GPU

Reply:
deleted it here and others.
I realized just now the fact that actually we only use nvcc, no standalone gcc. Then it's no need for guards of code that gcc cannot compile.